packages/gsul/src/getStatusUpdates.ts

```diff
@@ -0,0 +1,137 @@
+const tableName = process.env.TABLE_NAME
+
+const lambdaHandler = async (event: requestType): Promise<responseType> => {
+  // there are deliberately no try..catch blocks in this as any errors are caught by custom middy error handler
+  // and an error response is sent
+
+  const queryParams = event.prescriptions.map((prescription) => {
+    // create query for each prescription and ods code passed in
+    const queryParam : QueryCommandInput = {
+      TableName: tableName,
+      IndexName: "PharmacyODSCodePrescriptionIDIndex",
+      KeyConditionExpression: "PrescriptionID = :inputPrescriptionID AND PharmacyODSCode = :inputPharmacyODSCode",
+      ExpressionAttributeValues: {
+        ":inputPharmacyODSCode": prescription.odsCode,
+        ":inputPrescriptionID": prescription.prescriptionID
+      }
+    }
+    return queryParam
+  })
+
+  // run the dynamodb queries
+  const queryResultsTasks = runDynamoDBQueries(queryParams, docClient, logger)
+
+  // get all the query results
+  const queryResults = await Promise.all(queryResultsTasks)
+
+  // queryResults is an array of arrays
+  // the first array is for each prescription id passed in as we have created a query for each prescription
+  // and the second array is the items for that prescription id
+  // we use flat() to just get an array of items which we can pass to buildResults
+  const itemResults = buildResults(event.prescriptions, queryResults.flat())
+
+  const response = {
+    "schemaVersion": 1,
+    "isSuccess": true,
+    "prescriptions": itemResults
+  }
+  return response
+}
+
+export const buildResults = (inputPrescriptions: Array<inputPrescriptionType>,
+  queryResults: Array<DynamoDBResult>): Array<outputPrescriptionType> => {
+  // for each prescription id passed in build up a response
+  const itemResults = inputPrescriptions.map((prescription) => {
+    const items = queryResults.filter((queryResult) => {
+      // see if we have any results for the prescription id
+      return queryResult?.prescriptionID === prescription.prescriptionID
+    })
+    // if we have results then populate the response
+    if (items.length > 0) {
+      // get the latest update per item id
+
+      // first get the unique item ids
+      const uniqueItemIds = [...new Set(items.map((item) => {
+        return item.itemId
+      }))]
+
+      // now get an array of the latest updates for each unique item id
+      const latestUpdates = uniqueItemIds.map((itemId)=> {
+
+        // get all the updates for this prescription id and item id
+        const matchedItems = items.filter((item) => {
+          return item.prescriptionID === prescription.prescriptionID && item.itemId === itemId
+        })
+
+        // now just get the latest update
+        const latestUpdate = matchedItems.reduce((prev, current) => {
+          return (prev && Date.parse(prev.lastUpdateDateTime) > Date.parse(current.lastUpdateDateTime)) ? prev : current
+        })
+
+        return latestUpdate
+      })
```

Comment:
why not keep a dictionary, loop through each update, if the Id isn't in the dictionary add it, if it's already there compare dates and update or ignore as necessary.